schema/project.js

```diff
@@ -69,6 +70,75 @@ class ProjectQueryBuilder extends QueryBuilder {
     );
   }
 
+  whereRopsDue(year) {
+    return this.where('projects.issueDate', '<=', `${year}-12-31`)
+      .andWhere(builder => {
+        builder
+          .where('projects.status', 'active')
+          .orWhere(qb => {
+            qb
+              .where('projects.status', 'expired')
+              .where('projects.expiryDate', '>=', `${year}-01-01`);
+          })
+          .orWhere(qb => {
+            qb
+              .where('projects.status', 'revoked')
+              .where('projects.revocationDate', '>=', `${year}-01-01`);
+          });
+      });
+  }
+
+  whereRopsSubmitted(year) {
+    return this.whereExists(
+      Project.relatedQuery('rops')
+        .where('year', year)
+        .where('rops.status', 'submitted')
+    );
+  }
+
+  whereRopsOutstanding(year) {
+    return this.whereNotExists(
+      Project.relatedQuery('rops')
+        .where('year', year)
+        .where('rops.status', 'submitted')
+    );
+  }
+
+  withRops(year, ropsStatus) {
+    const query = this.withGraphFetched('rops(constrainRops)')
+      .modifiers({
+        constrainRops: builder => {
+          builder.where('rops.year', year);
+          if (ropsStatus === 'submitted') {
+            builder.where('rops.status', 'submitted');
+          }
+        }
+      });
+
+    if (ropsStatus === 'outstanding') {
+      const endOfJanNextYear = moment(`${parseInt(year, 10) + 1}-01-31`).endOf('day').toISOString();
+      query.select(this.knex().raw(`
+        CASE
+          WHEN projects.status = 'active' THEN '${endOfJanNextYear}'
+          WHEN projects.status = 'expired' THEN date_trunc('day', projects.expiry_date) + INTERVAL '29 days - 1 millisecond'
+          WHEN projects.status = 'revoked' THEN date_trunc('day', projects.revocation_date) + INTERVAL '29 days - 1 millisecond'
```

Comment:
This isn't _quite_ correct I'm afraid.

If a project is active, but expires before 3rd January then the ROP will be due at expiry + 28 days.

Reply:
You can probably achieve this with a `MIN(${endOfJanNextYear}, date_trunc('day', projects.expiry_date) + INTERVAL '29 days - 1 millisecond)`

Reply:
-1ms so that it expires at the end of the day, right?

Reply:
> -1ms so that it expires at the end of the day, right?

Yeah it gives us end of the 28th day.